Add Nix flake

VTimofeenko opened the pull request:
This commit embeds a Nix flake into the repository allowing a Nix user to run nitrocli by issuing:

    nix run d-e-s-o/nitrocli

This also adds a reproducible way to install nitrocli on a system controlled by Nix flakes.

robinkrahl commented:
Thanks for the PR!  Is it possible to add the man page and the shell completions to the flake?

VTimofeenko commented:
Sure. Would you prefer that to be in the same commit or a separate one?

d-e-s-o commented:
Apologies for the delay. I will should be able to take a closer look by the end of the weekend.

d-e-s-o commented on README.md:
````diff
@@ -112,6 +112,39 @@ $ cargo build --release
+
+Repository comes with a `flake.nix` file, so it can be run directly:
+
+``` bash
````
Seems `sh` is sufficient here, right?

d-e-s-o commented on flake.nix:
```diff
@@ -16,6 +16,17 @@
           nativeBuildInputs = with pkgs; [ hidapi ];
+          postInstall = ''
+            # copy the manpages
+            install -D -m 0644 ${./.}/doc/nitrocli.1 $out/share/man/man1
+            # make completions
+            mkdir -p $out/share/bash-completion/completions/
+            cargo run --bin=shell-complete bash > $out/share/bash-completion/completions/nitrocli
+            mkdir -p $out/share/zsh/site-functions/
+            cargo run --bin=shell-complete zsh > $out/share/zsh/site-functions/_nitrocli
+            mkdir -p $out/share/fish/vendor_completions.d/
+            cargo run --bin=shell-complete fish > $out/share/fish/vendor_completions.d/nitrocli.fish
```
Would you mind using long options when available? Makes it easier to understand the code.

d-e-s-o reviewed (commented):
> Sure. Would you prefer that to be in the same commit or a separate one?

Would you mind just creating a single commit, please? Commit #2 should almost definitely be in the initial commit of the flake, as otherwise it won't work to begin with (right?). We strive to have everything working at each commit. And I don't see a good reason to not contain the man page installation there as well.

I think the larger question is, though, should we actually check in these files here? For Gentoo, for example, they are maintained [out of tree](https://github.com/gentoo/gentoo/tree/master/app-crypt/nitrocli). I understand Ubuntu & Debian are conceptually similar (although all three support overlays of sorts, but getting accepted upstream significantly lowers barriers for users).
It's arguably a bit annoying (because there is no central place to find the recipes for all distros/systems/OSs), but at the same time we don't (tempted to say "can't, within reason") test them properly in the repository anyway, to the point that we have no control over certain aspects (for example a dependent package being renamed upstream) and, hence, could be prone to breakage over time. (I am also not a fan of dumping them into the repository root, but that's arguably a minor aspect)

Not being familiar with Nix, are flakes always managed in a decentralized manner? I found differing information in this respect. Could we, for example, aim to get the flake included [here](https://github.com/NixOS/nixpkgs) as a potentially easier to discover package source?

Can you please see if we can address the license check annotation failures?

VTimofeenko commented:
Hi!

>Would you mind just creating a single commit, please? 

Sure!

> Commit https://github.com/d-e-s-o/nitrocli/issues/2 should almost definitely be in the initial commit of the flake, as otherwise it won't work to begin with (right?). 

This commit only fixes development shell environment. Nix flakes allow to package the dependencies that are needed only for development in a separate development shell that the user can activate per project.

> We strive to have everything working at each commit. And I don't see a good reason to not contain the man page installation there as well.

Understood

> I think the larger question is, though, should we actually check in these files here? For Gentoo, for example, they are maintained [out of tree](https://github.com/gentoo/gentoo/tree/master/app-crypt/nitrocli). I understand Ubuntu & Debian are conceptually similar (although all three support overlays of sorts, but getting accepted upstream significantly lowers barriers for users).
It's arguably a bit annoying (because t[here](https://github.com/NixOS/nixpkgs) is no central place to find the recipes for all distros/systems/OSs), but at the same time we don't (tempted to say "can't, within reason") test them properly in the repository anyway, to the point that we have no control over certain aspects (for example a dependent package being renamed upstream) and, hence, could be prone to breakage over time. (I am also not a fan of dumping them into the repository root, but that's arguably a minor aspect)
> Not being familiar with Nix, are flakes always managed in a decentralized manner? I found differing information in this respect. Could we, for example, aim to get the flake included here as a potentially easier to discover package source?

Flakes are a bit special. TL;DR - flakes are explicitly a mechanism to embed the recipe to run a program into that program's repository

This is done so that a user of nix-the-package-manager can run on their system without installing that program system wide.

For example, without having neovim or any dependencies installed, I can run:

```
nix run "github:neovim/neovim?dir=contrib/"
```

in my terminal. Nix will resolve all dependencies and start neovim with all the dependencies and parameters specified in [flake.nix](https://github.com/neovim/neovim/blob/master/contrib/flake.nix). This will happen without altering the rest of my system.

> (I am also not a fan of dumping them into the repository root, but that's arguably a minor aspect)
I can move the files to a directory (similar to neovim) and adjust the README - just let me know. 

> Can you please see if we can address the license check annotation failures?

For `flake.nix` - definitely. `flake.lock` has to be json which does not allow for comments. Should I create a `flake.lock.json.license file` - same as `cargo.lock`?

d-e-s-o commented:
> > I think the larger question is, though, should we actually check in these files here? For Gentoo, for example, they are maintained [out of tree](https://github.com/gentoo/gentoo/tree/master/app-crypt/nitrocli). I understand Ubuntu & Debian are conceptually similar (although all three support overlays of sorts, but getting accepted upstream significantly lowers barriers for users).
> > It's arguably a bit annoying (because t[here](https://github.com/NixOS/nixpkgs) is no central place to find the recipes for all distros/systems/OSs), but at the same time we don't (tempted to say "can't, within reason") test them properly in the repository anyway, to the point that we have no control over certain aspects (for example a dependent package being renamed upstream) and, hence, could be prone to breakage over time. (I am also not a fan of dumping them into the repository root, but that's arguably a minor aspect)
> > Not being familiar with Nix, are flakes always managed in a decentralized manner? I found differing information in this respect. Could we, for example, aim to get the flake included here as a potentially easier to discover package source?
> 
> Flakes are a bit special. TL;DR - flakes are explicitly a mechanism to embed the recipe to run a program into that program's repository
> 
> This is done so that a user of nix-the-package-manager can run on their system without installing that program system wide.
> 
> For example, without having neovim or any dependencies installed, I can run:
> 
> ```
> nix run "github:neovim/neovim?dir=contrib/"
> ```
> 
> in my terminal. Nix will resolve all dependencies and start neovim with all the dependencies and parameters specified in [flake.nix](https://github.com/neovim/neovim/blob/master/contrib/flake.nix). This will happen without altering the rest of my system.

Thanks for the explanation; makes sense to me. For my understanding, though: What if went down the road of getting the flake into [`nixpkgs`](https://github.com/NixOS/nixpkgs) instead? The aforementioned command wouldn't get more complicated, I'd presume, but discoverability may be improved (if that's the "default" repository for people to look for flakes) and "they" may be better eqipped to maintain the recipe over time (for some definition of better).

I am not fundamentally opposed to the inclusion here, just want to make sure we have considered alternatives and ruled them out based on founded arguments.

> > (I am also not a fan of dumping them into the repository root, but that's arguably a minor aspect)
> > I can move the files to a directory (similar to neovim) and adjust the README - just let me know.

Perhaps we could move them into `pkg/nix`?

> > Can you please see if we can address the license check annotation failures?
> 
> For `flake.nix` - definitely. `flake.lock` has to be json which does not allow for comments. Should I create a `flake.lock.json.license file` - same as `cargo.lock`?

Yes, that sounds good to me. Probably ``flake.lock.license`, though.

VTimofeenko commented:
>Thanks for the explanation; makes sense to me. For my understanding, though: What if went down the road of getting the flake into [nixpkgs](https://github.com/NixOS/nixpkgs) instead? The aforementioned command wouldn't get more complicated, I'd presume, but discoverability may be improved (if that's the "default" repository for people to look for flakes) and "they" may be better eqipped to maintain the recipe over time (for some definition of better).

Nix ecosystem is notorious for overusing the term "nix" to mean several different things and generally being a bit confusing :)

To provide a parallel with python - nixpkgs is like pypi, and a flake is like `pyproject.toml` -  specific to a project or an environment and can contain the pinned dependencies, include development-only dependencies and tools configuration.

I am not that familiar with Rust, but if I understand correctly, pypi = crates.io and `pyproject.toml` = `Cargo.toml`.

In that sense - "having a flake in a project" is not conflicting with "project is available from nixpkgs". Flake approach is a bit more flexible, since the flake can be used in an environment that (for whatever reason) is pinned to an older version of nixpkgs without that project.

d-e-s-o commented:
Thanks for the explanation.

> In that sense - "having a flake in a project" is not conflicting with "project is available from nixpkgs".

Not conflicting, but it's a different set of trade-offs. The thing is, I don't really want to maintain a nix flake. I am not using it and can't do anything with it, to the degree that I cannot guarantee that it continues working the moment anything changes. And that latter aspect is fundamentally in conflict with what we intend to provide here in this repository. What should I do if a bug report comes in that the flake is no longer working? All I can do is throw my hands in the air and pray that somebody using Nix comes around and fixes the flake.

So from my perspective there are three ways to work around that:
1) Rely on `nixpkgs`. You would land the flake there and that is it (from my perspective). There is no association with this repository and so there is no assumption of folks here having ownership responsibilities. Maintenance falls on you or the next person requiring an updated version and submitting one. On top of that, discoverability may be improved, because (presumably) people can list the contents of `nixpkgs` and become aware of the program to begin with (whereas right now you basically have to know about it to reference it).
2) We have a CI recipe testing that builds on Nix. I don't know how involved that is. I'd think not too much, but haven't done research. We should look for precedent.
3) We strictly version any flakes provided within the repository and treat them as read-only once checked-in (barring potential bug fixes). E.g., you'd provide `nitrocli-0.4.1.nix` that builds `nitrocli` version `0.4.1`. Once `0.4.2` (or whatever) rolls around somebody would need to create an updated flake. `nitrocli-0.4.1.nix` may or may not stay around for a bit longer. In that sense it's similar to 1) (at least how I'd envision it; it's possible you'd have a flake that builds from `master` in `nixpkgs`, but I am not convinced that is wise). That is pretty much by far my least preferred variant, among other things because it (mis)uses this repository (which basically provides snapshots of development over time) as a datastore of sorts.

So after having thought about everything some more, I think that if we want to push through with the change as it is proposed currently, then we'd want CI integration. What are your thoughts?

VTimofeenko commented:
> The thing is, I don't really want to maintain a nix flake. I am not using it and can't do anything with it, to the degree that I cannot guarantee that it continues working the moment anything changes.

That's perfectly understandable, and I really do not wish to impose this contribution upon you and other maintainers :)

I would like to propose a fourth option - place it in a `contrib/` directory in this repository, and have a README.md there being very explicit about the nature of the flake (and, optionally, other future files in that directory). I believe the arguments for `contrib/` put forward by Drew DeVault in [this blog post](https://drewdevault.com/2020/06/06/Add-a-contrib-directory.html) are very sound.

From technical perspective, a flake is hermetic, it contains a pinned reference to nixpkgs, pinning the rest of the world required for it to run - so unless the nix-the-package-manager's flake interface changes, it will continue to function.

I am considering packaging nitrocli up for nixpkgs, but I am yet to familiarize myself with the way packages are created for nixpkgs - it's quite different from the flakes.

d-e-s-o commented:
> I would like to propose a fourth option - place it in a `contrib/` directory in this repository, and have a README.md there being very explicit about the nature of the flake (and, optionally, other future files in that directory). I believe the arguments for `contrib/` put forward by Drew DeVault in [this blog post](https://drewdevault.com/2020/06/06/Add-a-contrib-directory.html) are very sound.

On the flip side, to me at least it seems likely that such a directory may encourage sloppy practices, where now folks don't see the need to come up with a proper solution because their hack works for them and it can just be dumped there. Anyway, it's surely a trade-off.

I am not opposed to going down that route and trying it out in this specific case so feel free to adjust the pull request accordingly. That being said, if you integrate with `nixpkgs` down the line please give us a heads up so that we can remove the flake.

VTimofeenko commented:
I implemented the suggestions from the thread:

- Moved the flake to `contrib/nix`
- Added license information to flake.nix and flake.lock.license
- Adjusted README.md to include new path
- Squashed all commits into one

Sample output:

```
❯ nix run "git+file:///srv/code/forks/nitrocli?dir=contrib/nix" status
Status:
  model:             Nitrokey Storage
...
 ```

(git+file and the path are for my local fork)

Should I add `contrib/README.md` (and a README.md.license) with something along the lines of:

This is the `contrib/` directory of nitorcli project. Files located here:
- Are provided by the community of the project
- Are not official part of the project 
- Are provided without any warranty

?

d-e-s-o commented:
Thanks!

> Should I add `contrib/README.md` (and a README.md.license) with something along the lines of:

That would be great!

I hope to take a closer look this weekend.

d-e-s-o commented on contrib/nix/flake.nix:
```diff
@@ -0,0 +1,44 @@
+        };
+
+        devShell = with pkgs; mkShell {
+          buildInputs = [ cargo rustc rustfmt pre-commit rustPackages.clippy hidapi ];
```
For my understanding, why is `rustfmt` needed here?

VTimofeenko replied:
`rustfmt` just came from the [template](https://github.com/NixOS/templates/tree/master/rust).

> Is this file also suitable for installing all dependencies for development?

Correct, `devShell` contents can be triggered manually (`nix develop`) or through `direnv` to pull in development-only dependencies.

d-e-s-o commented:
Great stuff, thanks! I made some minor changes that hopefully didn't break anything (:o) and merged the patch.
